Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs

```diff
@@ -906,5 +956,130 @@ internal static TDevice Build<TDevice>(string layoutName = default, string layou
+                // update the state of any controls pointed to by the left node
+                var controlEndIndex = leftNode.controlStartIndex + leftNode.controlCount;
+                for (int i = leftNode.controlStartIndex; i < controlEndIndex; i++)
+                {
+                    var controlIndex = m_ControlTreeIndices[i];
+                    var control = m_ChildrenForEachControl[controlIndex];
+
+                    // nodes aren't always an exact fit for control memory ranges so check here if the control pointed
+                    // at by this node has actually changed state so we don't mark controls as stale needlessly.
+                    // We need to offset the device and new state pointers by the byte offset of the device state block
+                    // because all controls have this offset baked into them, but deviceStatePtr points at the already
+                    // offset block of device memory (remember, all devices share one big block of memory) and statePtr
+                    // points at a block of memory of the same size as the device state.
+                    if (!control.CompareState(deviceStatePtr - m_StateBlock.byteOffset, (byte*)statePtr - m_StateBlock.byteOffset, null))
+                        control.MarkAsStale();
+                }
+
+                // process the left child node if it exists
+                if (leftNode.leftChildIndex != -1)
+                    WriteChangedControlStatesInternal(statePtr, stateSizeInBits, deviceStatePtr,
+                        leftNode, startOffset);
+            }
+
+            // process the right child node if it exists
+            var rightNode = m_ControlTreeNodes[parentNode.leftChildIndex + 1];
+
+            // if no bits in the range defined by the right node have changed, return
+            if (!HasDataChangedInRange(deviceStatePtr, statePtr, leftNode.endBitOffset,
+                (uint)(rightNode.endBitOffset - leftNode.endBitOffset))) return;
+
+            // update the state of any controls pointed to by the right node
+            var rightNodeControlEndIndex = rightNode.controlStartIndex + rightNode.controlCount;
+            for (int i = rightNode.controlStartIndex; i < rightNodeControlEndIndex; i++)
+            {
+                var controlIndex = m_ControlTreeIndices[i];
+                var control = m_ChildrenForEachControl[controlIndex];
+
+                if (!control.CompareState(deviceStatePtr - m_StateBlock.byteOffset, (byte*)statePtr - m_StateBlock.byteOffset, null))
+                    control.MarkAsStale();
+            }
+
+            if (rightNode.leftChildIndex != -1)
+                WriteChangedControlStatesInternal(statePtr, stateSizeInBits, deviceStatePtr,
+                    rightNode, leftNode.endBitOffset);
+        }
+
+        private static unsafe bool HasDataChangedInRange(byte* deviceStatePtr, void* statePtr, uint startOffset, uint sizeInBits)
+        {
+            if (sizeInBits == 1)
```

Comment:
Is this branch necessary? Feels like MemCmp should contain single bit optimization instead